### DISCUSSION TITLE
Some notes on setting screen resolution

### DISCUSSION THREAD

**Discussion by justinclift** [notAperson535/OneClick-macOS-Simple-KVM#30] [General]
Hey all,

My setup has a wide screen monitor (3840x1600), which isn't easily supported by the existing resolution options in macOS.

So, after taking a few hours to get it working correctly (with success), I figured I might as well leave some potentially useful notes here.  Just in case someone else needs to do similar at some point. :smile:

## My setup

* Ryzen 5600X with a GTX 980 Ti graphics card
* Host OS is Ubuntu Linux 20.04
* The VM is (currently) running macOS Mojave, as I'm just needing the VM temporarily to track down a weird bug (that happens on Mojave) for a project I'm helping out
* The VM has been imported into Virt-Manager (libvirt), which probably affects some setting names

## Things I changed, to get it working

* Increased the graphics memory, assigned by Libvirt to the VM
  ```
  <video>
    <model type='qxl' ram='262144' vram='262144' vgamem='65536' heads='1' primary='yes'/>
    <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x0'/>
  </video>
  ```
  * I tried larger values first, but the VM wouldn't even boot.  So, I tried these... and things worked and I've left them there.

* Updated the `Resolution` tag in `EFI/EFI/OC/config.plist` (in the VM)
  ```
  <key>Resolution</key>
  <string>3840x1600@32</string>
  ```
  * There's a matching `ForceResolution` boolean (true or false value) a few entries above this in the file, that was useful when trying out video modes.  The final result (for my setup) needs that set to `false` for things to work.

* Had to compile a slightly customised version of TianoCore, with the `3840x1600` resolution I needed added in [here](https://github.com/tianocore/edk2/blob/ba0e0e4c6a174b71b18ccd6e47319cc45878893c/OvmfPkg/QemuVideoDxe/Initialize.c#L239).
  * Compiling this on Ubuntu 20.04 needs [Nasm](https://www.nasm.us/) 2.15.05 manually installed (`configure`, `make`, `make install`), as there's no Ubuntu package for this version of Nasm.  There is an older version of Nasm available for this Ubuntu release, but it errors out when trying to compile TianoCore.
  * When compiling TianoCore (using the [`edk2-stable202208`](https://github.com/tianocore/edk2/releases/tag/edk2-stable202208) release), these are the `Conf/target.txt` configuration file options I used:
    ```
    ACTIVE_PLATFORM       = OvmfPkg/OvmfPkgX64.dsc
    TARGET                = RELEASE
    TARGET_ARCH           = X64
    TOOL_CHAIN_TAG        = GCC5
    MAX_CONCURRENT_THREAD_NUMBER = 9
    ```
    * The other options in that file I just left at the defaults
  * When the compiling has finished (only takes about 1 minute), it generates a bunch of files into `Build/OvmfX64/RELEASE_GCC5/FV/` under the TianoCore source code directory.
    * The `OVMF_CODE.fd` and `OVMF_VARS.fd` ones are the ones I needed.  These need to be copied over the top of the existing `OVMF_CODE.fd` and `OVMF_VARS-1024x768.fd` files in the OneClick-macOS-Simple-KVM `firmware/` directory.
  * With those two updated  TianoCore files in place, I was then able to choose `3840x1600` from the list of resolutions in the UEFI boot screen, and macOS kept that resolution from then on.  Seems like a solid win (so far).

**Comment by notAperson535** [comment]
@justinclift please post the modified files here, I think it would benefit everyone.

**Comment by justinclift** [comment]
Ahhh, good idea. :smile:

These are the updated files: [firmware-3840x1600.zip](https://github.com/notAperson535/OneClick-macOS-Simple-KVM/files/9501995/firmware-3840x1600.zip)

* `OVMF_CODE.fd`, `OVMF_VARS.fd` : Updated Tianocore files, as output by the compiling process (above)
* `OVMF_VARS-3840x1600.fd` : The resulting OVMF_VARS.fd file, after I've gone through the UEFI dialog and chosen the new 3840x1600 resolution

I'll probably have the build environment still hanging around for a few weeks, so if someone wants me to add a few more resolutions and upload new files just let me know. :smile:

**Reply by notAperson535** [reply]
@justinclift two things, could you upload the `OVMF_VARS-1024x768.fd` file, and does the `OVMF_VARS-3840x1600.fd` change when you change resolutions?

**Reply by justinclift** [reply]
Errr, I don't have a `OVMF_VARS-1024x768.fd` file any more, as I renamed it to `OVMF_VARS-3840x1600.fd` (to avoid confusing myself, as it has the saved 3840x1600 resolution selected).

**Reply by justinclift** [reply]
> does the OVMF_VARS-3840x1600.fd change when you change resolutions?

Yep.  Just tested it now, and after changing the resolution from 3840x1600 to something else, that file is indeed changed.

**Reply by notAperson535** [reply]
@justinclift okay, thank you for the information!

**Reply by notAperson535** [reply]
@justinclift now try to get to that resolution using the files only from my repo, and tell me if I have fixed the issue.

**Reply by justinclift** [reply]
Good thinking.  Just tried that now (blowing away my firmware directory, then downloading the files from this repo).

The 3840x1600 resolution is indeed in the list, and selecting it + saving the change works properly.

![2022 09 08_1-kvm_tianocore_has_correct_resolution_available](https://user-images.githubusercontent.com/406299/189115897-d2649b6d-d7e8-46d4-9857-04c8dbe3aee9.png)

The new setting is correctly kept.  Screenshot below is after turning off the vm, then starting it again and going into the UEFI menu to check:

![2022 09 08_2-the_resolution_persists_between_restarts_too](https://user-images.githubusercontent.com/406299/189115912-92269f69-d286-4612-bbae-0e81450b00d7.png)

**Reply by notAperson535** [reply]
@justinclift great!

**Comment by justinclift** [comment]
This is my updated Libvirt domain xml file too (with sanitised path names), in case that's useful as well: [libvirt_macos.zip](https://github.com/notAperson535/OneClick-macOS-Simple-KVM/files/9502094/libvirt_macos.zip)

**Comment by justinclift** [comment]
As a note for anyone needing to add further resolutions to the TianoCore file, these are the commands needed for the EDKII compile process:

```
$ make -C BaseTools
$ . edksetup.sh  <-- this sets environment variables to find the "build" command below
$ nano Conf/target.txt   <-- change the values to match the top post in this discussion
$ build
```

The "How to compile this stuff" documentation I followed originally is here:

https://github.com/tianocore/tianocore.github.io/wiki/Common-instructions

The EDKII project seems to have moved to a new compile approach since then, but the above process is the one I used and still works fine.

**Comment by justinclift** [comment]
In theory, this firmware should have `2560x1080` resolution now included as well:

[firmware-2560x1080.zip](https://github.com/notAperson535/OneClick-macOS-Simple-KVM/files/14895354/firmware-2560x1080.zip)

Inside that .zip file is `OVMF_CODE.fd` which has been updated to also include the new 2560x1080 resolution.  You'll need to shut down your macOS vm, replace your existing `OVMF_CODE.fd` with this new one, then start up your vm again + go into the early boot BIOS like boot settings to choose the new resolution.

**Comment by justinclift** [comment]
@notAperson535 This is the newer TianoCore 2024.02 release (compiled it just now), with the two extra video modes added:

* 2560x1080
* 3840x1600

[firmware-edk2-stable202402.zip](https://github.com/notAperson535/OneClick-macOS-Simple-KVM/files/14899170/firmware-edk2-stable202402.zip)

It's completely untested though (!), so better to not blindly assume it's all good.

The `OVMF_VARS.fd` file in there is in the default state that pops out of the compile process, so you might need to update some defaults in it (screen resolution?) to be suitable for others to use.

Hopefully that helps. :smile:

**Comment by davidricardo1026** [comment]
thanks a lot ,but with this 
```xml
  <os>
    <type arch="x86_64" machine="pc-q35-4.2">hvm</type>
    <loader readonly="yes" type="pflash" format="raw">/home/ss/Downloads/source/firmware/OVMF_CODE.fd</loader>
    <nvram format="raw">/home/ss/Downloads/source/firmware/OVMF_VARS-3840x1600.fd</nvram>
  </os>
```

i still just only can check 1920*1080, the last items lost
here is my [gist](https://gist.github.com/davidricardo1026/6ab394b63b4f67c77ef5d0c4bfb2613a)

**Reply by davidricardo1026** [reply]
> thanks a lot ,but with this
> 
> ```
>   <os>
>     <type arch="x86_64" machine="pc-q35-4.2">hvm</type>
>     <loader readonly="yes" type="pflash" format="raw">/home/ss/Downloads/source/firmware/OVMF_CODE.fd</loader>
>     <nvram format="raw">/home/ss/Downloads/source/firmware/OVMF_VARS-3840x1600.fd</nvram>
>   </os>
> ```
> 
> i still just only can check 1920*1080, the last items lost here is my [gist](https://gist.github.com/davidricardo1026/6ab394b63b4f67c77ef5d0c4bfb2613a)

OSX-KVM,my error, [OSX-KVM](https://github.com/kholia/OSX-KVM) is too old ,and just can only run
yours is the best.
follow yours is the best and no error
